### PR TITLE
[clangd] Disable crashy unchecked-optional-access tidy check

### LINKED ISSUE #1700: clangd stop in "parsing includes, running Build AST" for long time becase of bugprone-unchecked-optional-access
I found that Clangd often gets stuck in the 'parsing includes, running Build AST', and after using Mac's profile tool, I found that the actual stack is as follows

image

after disable `bugprone-unchecked-optional-access` in clangd configuration, every thing be ok.

I wonder if we can separately prompt for 'runing clang-tidy check xxxxx' instead of displaying 'parsing includes, running Build AST'

image

> It is difficult to provide a minimum recurrence case because this problem occurs under a mountain of shit😂

### LINKED ISSUE #69369: dead loop in clang::dataflow::WatchedLiteralsSolverImpl::solve(long long) when running bugprone-unchecked-optional-access tidy-check
Reproduce step:
1. build https://github.com/ccache/ccache with compile_commands.json
2. with 17.02 clangd check src/ccache.cpp

Stack trace
```
 thread #1, queue = 'com.apple.main-thread', stop reason = instruction step over
  * frame #0: 0x00000001063b1a44 clangd`clang::dataflow::WatchedLiteralsSolverImpl::solve(long long) && + 228
    frame #1: 0x00000001063b17a4 clangd`clang::dataflow::WatchedLiteralsSolver::solve(llvm::ArrayRef<clang::dataflow::Formula const*>) + 64
    frame #2: 0x000000010639bc84 clangd`clang::dataflow::DataflowAnalysisContext::isUnsatisfiable(llvm::SetVector<clang::dataflow::Formula const*, std::__1::vector<clang::dataflow::Formula const*, std::__1::allocator<clang::dataflow::Formula const*>>, llvm::DenseSet<clang::dataflow::Formula const*, llvm::DenseMapInfo<clang::dataflow::Formula const*, void>>, 0u>) + 180
    frame #3: 0x000000010639b8d8 clangd`clang::dataflow::DataflowAnalysisContext::flowConditionImplies(clang::dataflow::Atom, clang::dataflow::Formula const&) + 196
    frame #4: 0x000000010638853c clangd`clang::dataflow::UncheckedOptionalAccessModel::compare(clang::QualType, clang::dataflow::Value const&, clang::dataflow::Environment const&, clang::dataflow::Value const&, clang::dataflow::Environment const&) + 280
    frame #5: 0x0000000106388978 clangd`clang::dataflow::UncheckedOptionalAccessModel::widen(clang::QualType, clang::dataflow::Value&, clang::dataflow::Environment const&, clang::dataflow::Value&, clang::dataflow::Environment&) + 44
    frame #6: 0x00000001063a1588 clangd`clang::dataflow::Environment::widen(clang::dataflow::Environment const&, clang::dataflow::Environment::ValueModel&) + 404
    frame #7: 0x00000001063aee8c clangd`clang::dataflow::runTypeErasedDataflowAnalysis(clang::dataflow::ControlFlowContext const&, clang::dataflow::TypeErasedDataflowAnalysis&, clang::dataflow::Environment const&, std::__1::function<void (clang::CFGElement const&, clang::dataflow::TypeErasedDataflowAnalysisState const&)>) + 1340
    frame #8: 0x000000010635f5b4 clangd`llvm::Expected<std::__1::vector<std::__1::optional<clang::dataflow::DataflowAnalysisState<clang::dataflow::UncheckedOptionalAccessModel::Lattice>>, std::__1::allocator<std::__1::optional<clang::dataflow::DataflowAnalysisState<clang::dataflow::UncheckedOptionalAccessModel::Lattice>>>>> clang::dataflow::runDataflowAnalysis<clang::dataflow::UncheckedOptionalAccessModel>(clang::dataflow::ControlFlowContext const&, clang::dataflow::UncheckedOptionalAccessModel&, clang::dataflow::Environment const&, std::__1::function<void (clang::CFGElement const&, clang::dataflow::DataflowAnalysisState<clang::dataflow::UncheckedOptionalAccessModel::Lattice> const&)>) + 216
    frame #9: 0x000000010635f13c clangd`clang::tidy::bugprone::UncheckedOptionalAccessCheck::check(clang::ast_matchers::MatchFinder::MatchResult const&) + 464
    frame #10: 0x0000000106c42058 clangd`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::MatchVisitor::visitMatch(clang::ast_matchers::BoundNodes const&) + 132
    frame #11: 0x0000000106c6d1a0 clangd`clang::ast_matchers::internal::BoundNodesTreeBuilder::visitMatches(clang::ast_matchers::internal::BoundNodesTreeBuilder::Visitor*) + 256
    frame #12: 0x0000000106c419ac clangd`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::matchWithFilter(clang::DynTypedNode const&) + 1104
    frame #13: 0x0000000106c433fc clangd`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::TraverseDecl(clang::Decl*) + 276
    frame #14: 0x0000000106c46728 clangd`clang::RecursiveASTVisitor<clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor>::TraverseDecl(clang::Decl*) + 13004
    frame #15: 0x0000000106c43408 clangd`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::TraverseDecl(clang::Decl*) + 288
    frame #16: 0x0000000106c226bc clangd`clang::ast_matchers::MatchFinder::matchAST(clang::ASTContext&) + 748
    frame #17: 0x0000000105fca154 clangd`clang::clangd::ParsedAST::build(llvm::StringRef, clang::clangd::ParseInputs const&, std::__1::unique_ptr<clang::CompilerInvocation, std::__1::default_delete<clang::CompilerInvocation>>, llvm::ArrayRef<clang::clangd::Diag>, std::__1::shared_ptr<clang::clangd::PreambleData const>) + 7508
    frame #18: 0x0000000105e63994 clangd`clang::clangd::check(llvm::StringRef, clang::clangd::ThreadsafeFS const&, clang::clangd::ClangdLSPServer::Options const&) + 7376
    frame #19: 0x0000000105e53294 clangd`clang::clangd::clangdMain(int, char**) + 8664
    frame #20: 0x000000018659bf28 dyld`start + 2236
```

### PR DESCRIPTION
**Pull request by kadircet**
Fixes https://github.com/llvm/llvm-project/issues/69369.
Fixes https://github.com/clangd/clangd/issues/1700.

### CODE REVIEW

File: clang-tools-extra/clangd/TidyProvider.cpp [hunk 1]
```diff
@@ -219,6 +219,9 @@ TidyProvider disableUnusableChecks(llvm::ArrayRef<std::string> ExtraBadChecks) {
       "-bugprone-use-after-move",
       // Alias for bugprone-use-after-move.
       "-hicpp-invalid-access-moved",
+      // Check uses dataflow analysis, which might hang/crash unexpectedly on
+      // incomplete code.
+      "-bugprone-unchecked-optional-access",
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/TidyProvider.cpp:224]: A consequence of putting the check here, is that a user who wants the check in spite of its bugginess does not have a way to enable it.

It was suggested, in https://github.com/clangd/clangd/issues/1700#issuecomment-1659690492 and https://github.com/clangd/clangd/issues/1337#issuecomment-1610876189, that we should have a way to allow users to override the disablements in this "bad checks" list. Could you weigh in on that proposal?

Comment (kadircet) [reply clang-tools-extra/clangd/TidyProvider.cpp:224]: FWIW, we had this check enabled internally before it was upstream'd and experience on incomplete code has always been problematic. Hence I don't see much point in letting people run this on an incomplete AST. I believe the recommended workflow should be through running clang-tidy binary directly. I'll add some comments on those threads too.

**Review by HighCommander4 — APPROVED** [review]

**Comment by HighCommander4** [comment]
This may be a good candidate for a 17.0.x backport.

**Comment by kadircet** [comment]
yeah, initiated that (i think) already in https://github.com/llvm/llvm-project/issues/69568

**Pull request merged**